### PR TITLE
Add support for 'one-shot' services using NoRestart k8s Jobs

### PR DESCRIPTION
**Pull request by drshrey**

### CODE REVIEW

File: server/plugins/kubedeploy/deployments.go [hunk 1]
```diff
@@ -139,7 +140,7 @@ func (x *KubeDeploy) createNamespaceIfNotExists(namespace string, coreInterface
 
 func (x *KubeDeploy) doDeploy(e agent.Event) error {
 	// Codeflow will load the kube config from a file, specified by CF_PLUGINS_KUBEDEPLOY_KUBECONFIG environment variable
-	kubeconfig := viper.GetString("plugins.kubedeploy.kubeconfig")
+	kubeconfig := "/Users/shreyas/.kube/config"
```
Comment (sasso) [inline server/plugins/kubedeploy/deployments.go]: you should change that back to `viper.GetString("plugins.kubedeploy.kubeconfig")` and use `export CF_PLUGINS_KUBEDEPLOY_KUBECONFIG=/Users/shreyas/.kube/config`

File: server/plugins/kubedeploy/deployments.go [hunk 2]
```diff
@@ -391,90 +393,155 @@ func (x *KubeDeploy) doDeploy(e agent.Event) error {
...
+			jobParams = &apis_batch_v1.Job{
+				TypeMeta: unversioned.TypeMeta{
+					Kind:       "Job",
+					APIVersion: "batch/v1",
+				},
+				ObjectMeta: v1.ObjectMeta{
+					Name: deploymentName,
+				},
+				Spec: apis_batch_v1.JobSpec{
+					Parallelism: &numParallelPods,
```
Comment (sasso) [inline server/plugins/kubedeploy/deployments.go]: You could use `util.Int32Ptr` here

File: server/plugins/kubedeploy/deployments.go [hunk 3]
```diff
@@ -391,90 +393,155 @@ func (x *KubeDeploy) doDeploy(e agent.Event) error {
...
+
+				_, myError = batchv1DepInterface.Jobs(namespace).Create(jobParams)
+
+				if myError != nil {
+					// send failed status
+					log.Printf("Failed to create service job %s, with error: %s", deploymentName, myError)
+					data.Services[index].State = plugins.Failed
+					data.Services[index].StateMessage = fmt.Sprintf("Error creating deployment: %s", myError)
+					// shorten the timeout in this case so that we can fail without waiting
+					curTime = timeout
```
Comment (sasso) [inline server/plugins/kubedeploy/deployments.go]: Could this timeout break the one shot job? Let's say we have a migration that will take 1h?

File: server/plugins/kubedeploy/deployments_test.go [hunk 4]
```diff
@@ -1,11 +1,13 @@
 package kubedeploy_test
 
 import (
+	"log"
 	"testing"
 
 	"github.com/checkr/codeflow/server/agent"
 	"github.com/checkr/codeflow/server/plugins"
 	"github.com/checkr/codeflow/server/plugins/kubedeploy/testdata"
+	"github.com/davecgh/go-spew/spew"
```
Comment (sasso) [inline server/plugins/kubedeploy/deployments_test.go]: Do we need spew here?

File: server/plugins/kubedeploy/deployments.go [hunk 5]
```diff
@@ -287,20 +308,173 @@ func (x *KubeDeploy) doDeploy(e agent.Event) error {
...
+		}
+	}
+
+	for index, service := range oneShotServices {
+		oneShotServiceName := strings.ToLower(genOneShortServiceName(data.Project.Slug, service.Name))
+
+		// Check and delete any completed or failed jobs, and delete respective pods
+		existingJobs, err := batchv1DepInterface.Jobs(namespace).List(v1.ListOptions{LabelSelector: fmt.Sprintf("%s=%s", "app", oneShotServiceName)})
+		if err != nil {
+			log.Printf("Failed to load existing jobs with label app=%s, with error: %s", oneShotServiceName, err)
```
Comment (jeremyd) [inline server/plugins/kubedeploy/deployments.go]: This message that we log should be the same as the one sent in the Error response.  This error would be failure to list jobs... (and not failed to create job).

File: server/plugins/kubedeploy/deployments.go [hunk 6]
```diff
@@ -287,20 +308,173 @@ func (x *KubeDeploy) doDeploy(e agent.Event) error {
...
-			deployPorts = append(deployPorts, newContainerPort)
+
+			// delete old job
+			gracePeriod := int64(0)
+			isOrphan := true
+			deleteOptions := v1.DeleteOptions{
+				GracePeriodSeconds: &gracePeriod,
+				OrphanDependents:   &isOrphan,
+			}
+			batchv1DepInterface.Jobs(namespace).Delete(job.Name, &deleteOptions)
```
Comment (jeremyd) [inline server/plugins/kubedeploy/deployments.go]: Lets log the error if this Delete call fails.

File: server/plugins/kubedeploy/deployments.go [hunk 7]
```diff
@@ -287,20 +308,173 @@ func (x *KubeDeploy) doDeploy(e agent.Event) error {
...
+				} else {
+					// Job has failed!
+					oneShotServices[index].State = plugins.Failed
+					oneShotServices[index].StateMessage = fmt.Sprintf("Error job has failed %s", oneShotServiceName)
+					x.sendDDErrorResponse(e, oneShotServices, oneShotServices[index].StateMessage)
+					return nil
+				}
+			}
+
+			time.Sleep(5 * time.Second)
```
Comment (jeremyd) [inline server/plugins/kubedeploy/deployments.go]: would this line of code ever execute?

File: server/plugins/kubedeploy/deployments.go (hunk 6 above)
Comment (drshrey) [reply server/plugins/kubedeploy/deployments.go]: done

File: server/plugins/kubedeploy/deployments.go (hunk 5 above)
Comment (drshrey) [reply server/plugins/kubedeploy/deployments.go]: done

**Comment by drshrey** [comment]
Continued addressing in https://github.com/checkr/codeflow/pull/148

**Pull request closed without merging**